drivers/bluetooth/hci/hci_nxp.c

```diff
@@ -0,0 +1,1178 @@
+	uart_dev_data.primary_speed = DT_PROP_OR(HCI_UART_CHILD, fw_download_primary_speed, speed);
+	uart_dev_data.secondary_speed =
+		DT_PROP_OR(HCI_UART_CHILD, fw_download_secondary_speed, speed);
+
+	flowcontrol_of_hci = DT_PROP_OR(HCI_UART_CHILD, hw_flow_control, false) ? true : false;
+	uart_dev_data.primary_flowcontrol =
+		DT_PROP_OR(HCI_UART_CHILD, fw_download_primary_flowcontrol, false) ? true : false;
+	uart_dev_data.secondary_flowcontrol =
+		DT_PROP_OR(HCI_UART_CHILD, fw_download_secondary_flowcontrol, false) ? true : false;
+
+#if DT_NODE_HAS_PROP(HCI_UART_CHILD, sdio_reset_gpios) ||                                          \
+	DT_NODE_HAS_PROP(HCI_UART_CHILD, w_disable_gpios)
+#if DT_NODE_HAS_PROP(HCI_UART_CHILD, sdio_reset_gpios)
+	/* Check BT REG_ON gpio instance */
+	if (!gpio_is_ready_dt(&sido_reset)) {
+		LOG_ERR("Error: failed to configure sido_reset %s pin %d", sido_reset.port->name,
+			sido_reset.pin);
+		return -EIO;
+	}
+
+	/* Configure sido_reset as output  */
+	err = gpio_pin_configure_dt(&sido_reset, GPIO_OUTPUT);
+	if (err) {
+		LOG_ERR("Error %d: failed to configure sido_reset %s pin %d", err,
+			sido_reset.port->name, sido_reset.pin);
+		return err;
+	}
+	err = gpio_pin_set_dt(&sido_reset, 0);
+	if (err) {
+		return err;
+	}
+#endif /* DT_NODE_HAS_PROP(HCI_UART_CHILD, sdio_reset_gpios) */
+
+#if DT_NODE_HAS_PROP(HCI_UART_CHILD, w_disable_gpios)
+	/* Check BT REG_ON gpio instance */
+	if (!gpio_is_ready_dt(&w_disable)) {
+		LOG_ERR("Error: failed to configure w_disable %s pin %d", w_disable.port->name,
+			w_disable.pin);
+		return -EIO;
+	}
+
+	/* Configure w_disable as output  */
+	err = gpio_pin_configure_dt(&w_disable, GPIO_OUTPUT);
+	if (err) {
+		LOG_ERR("Error %d: failed to configure w_disable %s pin %d", err,
+			w_disable.port->name, w_disable.pin);
+		return err;
+	}
+	err = gpio_pin_set_dt(&w_disable, 0);
+	if (err) {
+		return err;
+	}
+#endif /* DT_NODE_HAS_PROP(HCI_UART_CHILD, w_disable_gpios) */
+
+	/* wait for reset done */
+	k_sleep(K_MSEC(100));
+
+#if DT_NODE_HAS_PROP(HCI_UART_CHILD, sdio_reset_gpios)
+	err = gpio_pin_set_dt(&sido_reset, 1);
+	if (err) {
+		return err;
+	}
+#endif /* DT_NODE_HAS_PROP(HCI_UART_CHILD, sdio_reset_gpios) */
+
+#if DT_NODE_HAS_PROP(HCI_UART_CHILD, w_disable_gpios)
+	err = gpio_pin_set_dt(&w_disable, 1);
+	if (err) {
+		return err;
+	}
+#endif /* DT_NODE_HAS_PROP(HCI_UART_CHILD, w_disable_gpios) */
+#endif
+
+	uart_irq_rx_disable(uart_dev);
+	uart_irq_tx_disable(uart_dev);
+
+	fw_upload.rx.head = 0;
+	fw_upload.rx.tail = 0;
+
+	k_sem_init(&fw_upload.rx.sem, 0, sizeof(fw_upload.rx.buffer));
+
+	uart_irq_callback_set(uart_dev, bt_nxp_ctlr_uart_isr);
+
+	made_table = false;
+
+	fw_upload_initCrc8();
+
+	err = fw_upload_uart_reconfig(uart_dev_data.primary_speed,
+				      uart_dev_data.primary_flowcontrol);
+	if (err) {
+		LOG_ERR("Fail to config uart");
+		return err;
+	}
+
+	uart_irq_rx_enable(uart_dev);
+
+	err = fw_uploading(bt_fw_bin, bt_fw_bin_len);
```

Comment:
@lylezhu2012, We need to put together a request to the project to handle the binary blob for this firmware. I'll take this offline with you to get this setup.

Reply:
This PR is used to implement a driver to download controller firmware via UART.
I have a guide here https://github.com/zephyrproject-rtos/zephyr/pull/70815/commits/7f5f360ec1060f5eba55d3d009ee7fa3ebb76a10#diff-dccbb25cc34086042193126bd5b09e2b1ff09d3f24fb18da73095b6c2244560cR252-R275 to show how to download firmware.

Once the binary blob is done, I will update the driver and guide.